apps/dhparam.c

```diff
@@ -281,49 +241,70 @@ int dhparam_main(int argc, char **argv)
                  * We have no PEM header to determine what type of DH params it
                  * is. We'll just try both.
                  */
-                dh = alloc_dh = ASN1_d2i_bio_of(DH, DH_new, d2i_DHparams, in, NULL);
+                params = d2i_KeyParams_bio(EVP_PKEY_DH, NULL, in);
                 /* BIO_reset() returns 0 for success for file BIOs only!!! */
-                if (dh == NULL && BIO_reset(in) == 0)
-                    dh = alloc_dh = ASN1_d2i_bio_of(DH, DH_new, d2i_DHxparams, in, NULL);
+                if (params == NULL && BIO_reset(in) == 0)
+                    params = d2i_KeyParams_bio(EVP_PKEY_DHX, NULL, in);
             } else {
                 /* informat == FORMAT_PEM */
-                dh = alloc_dh = PEM_read_bio_DHparams(in, NULL, NULL, NULL);
+                params = PEM_read_bio_Parameters(in, NULL);
             }
 
-            if (dh == NULL) {
+            if (params == NULL
+                || (EVP_PKEY_base_id(params) != EVP_PKEY_DH
+                    && EVP_PKEY_base_id(params) != EVP_PKEY_DHX)) {
+                EVP_PKEY_free(params);
                 BIO_printf(bio_err, "Error, unable to load DH parameters\n");
                 goto end;
             }
         }
-        /* dh != NULL */
     }
 
+#if !defined(OPENSSL_NO_DSA) && !defined(OPENSSL_NO_DEPRECATED_3_0)
+    if (dsaparam) {
+        DSA *dsa;
+        DH *dh;
+
+        dsa = EVP_PKEY_get0_DSA(params);
+        if (dsa == NULL)
+            goto end;
+        dh = DSA_dup_DH(dsa);
```

Comment:
This is worth a `NULL` check although it will not cause problem without the check - but it will result in unnecessary function calls below.